spouse/spouse_demo.py

```diff
@@ -0,0 +1,363 @@
+#
+# Make sure `dbpedia.pkl` is in the `spouse/data` directory.
+
+# %%
+with open("data/dbpedia.pkl", "rb") as f:
+    known_spouses = pickle.load(f)
+
+list(known_spouses)[0:5]
+
+
+# %%
+@labeling_function(resources=dict(known_spouses=known_spouses), pre=[get_person_text])
+def lf_distant_supervision(x, known_spouses):
+    p1, p2 = x.person_names
+    if (p1, p2) in known_spouses or (p2, p1) in known_spouses:
+        return POSITIVE
+    else:
+        return ABSTAIN
+
+
+# %%
+from preprocessors import last_name
+
+# Last name pairs for known spouses
+last_names = set(
+    [
+        (last_name(x), last_name(y))
+        for x, y in known_spouses
+        if last_name(x) and last_name(y)
+    ]
+)
+
+
+@labeling_function(resources=dict(last_names=last_names), pre=[get_person_last_names])
+def lf_distant_supervision_last_names(x, last_names):
+    p1_ln, p2_ln = x.person_lastnames
+
+    return (
+        POSITIVE
+        if (p1_ln != p2_ln)
+        and ((p1_ln, p2_ln) in last_names or (p2_ln, p1_ln) in last_names)
+        else ABSTAIN
+    )
+
+
+# %% [markdown]
+# #### Apply Labeling Functions to the Data
+# We create a list of labeling functions and apply them to the data
+
+# %%
+from snorkel.labeling.apply import PandasLFApplier
+
+lfs = [
+    lf_husband_wife,
+    lf_husband_wife_left_window,
+    lf_same_last_name,
+    lf_married,
+    lf_familial_relationship,
+    lf_family_left_window,
+    lf_other_relationship,
+    lf_distant_supervision,
+    lf_distant_supervision_last_names,
+]
+applier = PandasLFApplier(lfs)
+
+# %%
+from snorkel.labeling.analysis import LFAnalysis
+
+dev_L = applier.apply(df_dev)
+train_L = applier.apply(df_train)
+
+LFAnalysis(dev_L, lfs).lf_summary(Y_dev)
+
+# %% [markdown]
+# ### Training the Label Model
+#
+# Now, we'll train a model of the LFs to estimate their weights and combine their outputs. Once the model is trained, we can combine the outputs of the LFs into a single, noise-aware training label set for our extractor.
+
+# %%
+from snorkel.labeling.model.label_model import LabelModel
+
+label_model = LabelModel(cardinality=2, verbose=True)
+label_model.fit(train_L, Y_dev, n_epochs=5000, log_freq=500, seed=12345)
+
+# %% [markdown]
+# ### Label Model Metrics
+# Since our dataset is highly unbalanced (91% of the labels are negative), even a trivial baseline that always outputs negative can get a high accuracy. So we evaluate the label model using the F1 score and ROC-AUC rather than accuracy.
+
+# %%
+from snorkel.analysis.metrics import metric_score
+from snorkel.analysis.utils import probs_to_preds
+
+Y_probs_dev = label_model.predict_proba(dev_L)
+Y_preds_dev = probs_to_preds(Y_probs_dev)
+print(
+    f"Label model f1 score: {metric_score(Y_dev, Y_preds_dev, probs=Y_probs_dev, metric='f1')}"
+)
+print(
+    f"Label model roc-auc: {metric_score(Y_dev, Y_preds_dev, probs=Y_probs_dev, metric='roc_auc')}"
+)
+
+# %%
+import matplotlib.pyplot as plt
```

Comment:
Can we add a description about why we are showing this plot?

Reply:
Removed it. I had already removed the explanatory text earlier but forgot to remove this.